anthony/unrealized-potential-of-string-diagrams.md

```diff
@@ -0,0 +1,128 @@
+
+There are, of course, many clever ways around this. I was recently pointed to this as a way to arithmetize combinator calculi in a different way than I suggest.
+
+- [EDEN - a practical, SNARK-friendly combinator VM and ISA](https://eprint.iacr.org/2023/1021.pdf)
+
+Based on my understanding, it's using a combinatorial construction which exploits an isomorphism between tree expressions and Dyck words to linearize tree expressions in a way that's more SNARK friendly. I think there is a lot of potential in using latice paths, pattern-avoiding permutations, and other linear combinatorial objects to create effective arithmetizations for higher-dimensional combinatorial structures. I don't want to give the impression that this post is about "the one true way". I think it's better than alternatives I've seen, but there's so much low hanging fruit all over the design space that's worth exploring.
+
+Finding a good arithmetization can be cast as the problem of fitting your check into as thin a matrix as possible. We have a structure, $D$, that we want to check with a predicate, $P$. We want to create a function, $f$, that turns any given $D$ into a matrix, and devise a predicate over matricies, $P'$, such that $P'(m) \equiv (\exists d. P(d) \wedge f(d) = m)$. Ideally, the width of $f(d)$ should be $O(1)$ with respect to the size of $d$. Of course, we also want $P'$ to be equivalent to a low-degree polynomial equation, but I'm going to ignore that for now.
+
+Let's get a bit more technical about what a string diagram is. In its simplest form, we have a list of nodes, each having a list of ports. The wiring can be characterized in a few ways. Often, the ports will be partitioned into "input" and "output" (or "positive" and "negative") and the wiring is just a bijection between the input and output ports. Without such a partition, we can instead characterize the wiring as an automorphism over the ports such that no port is mapped onto itself. String diagrams will often have additional structure, but this is the basic idea. This captures string diagrams as a streightforward combinatorial object.
+
+A rewrite rule consists of a pair of sub-graphs wired into a boundary. This boundary is shared between both sides of the rewrite rule. We can define a rewrite rule as a triple consisting of a list of input nodes, a list of output nodes, a boundary with its own list of ports, and two automorphisms; one over the sum of input and boundary ports and the other over output and boundary ports. We may also have an additional requirements, such as having no boundary port be mapped onto another boundary port.
```

Comment:
on rewriting: there's been work by Pawel, e.g., 
https://dl.acm.org/doi/abs/10.1145/3502719?download=true

and some using DPO approach to graph rewriting